File: tic_tac_toe/tic_tac_toe.py
#!/usr/bin/python3
import random

# mr tic tac toe
mrttt = '''
       XXXXXXXXXXXXXXXXXXXXXX
    XXXXXXXXXXXXXXXXXXXXXXXXXXXXX
  XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
 XXXXXXXXXXXXXXXXXX         XXXXXXXX
XXXXXXXXXXXXXXXX              XXXXXXX
XXXXXXXXXXXXX                   XXXXX
 XXX     _________ _________     XXX      
  XX    I  _xxxxx I xxxxx_  I    XX        
 ( X----I         I         I----X )           
( +I    I      00 I 00      I    I+ )
 ( I    I    __0  I  0__    I    I )      /\    /\    |""")  -------  -------  -------
  (I    I______ /   \_______I    I)      /  \  /  \   |___)     |   __   |   __   |
   I           ( ___ )           I      /    \/    \  |   \ *   |        |        |
   I    _  :::::::::::::::  _    I           
    \    \___ ::::::::: ___/    /          
     \_      \_________/      _/
       \        \___/        /
         \                 /
          |\             /|
          |  \_________/  |
 \n'''
# values of board and print the board
bval = [0, 1, 2, 3, 4, 5, 6, 7, 8]


def board():
    print("\n")
    print("     |     |")
    print("  {}  |  {}  |  {}".format(bval[0], bval[1], bval[2]))
    print("_____|_____|_____")
    print("     |     |")
    print("  {}  |  {}  |  {}".format(bval[3], bval[4], bval[5]))
    print("_____|_____|_____")
    print("     |     |")
    print("  {}  |  {}  |  {}".format(bval[6], bval[7], bval[8]))
    print("     |     |")
    print("\n")


# make the menu input name and X/O choice give the other choice to cpu
def menu():
    i = 0
    print(mrttt)
    print("WELCOME PLAYER!!! I am Mr. Tic Tac Toe \nShall we play some TIC TAC TOE? \n")
    playername = input("What is your name challenger ?\n")
    while i == 0:
        try:
            xno = int(
                input(
                    "{},Select your mark: X or O \nFor X Press 1 \nFor O Press 0 \n".format(
                        playername
                    )
                )
            )
        except ValueError:
            print("\nWrong Input!!! Try Again \n")
            continue
        if xno == 1 or xno == 0:
            i = 1
            if xno == 1:
                human = "X"
                cpu = "O"
            else:
                human = "O"
                cpu = "X"
        else:
            print("\n Wrong input! Try Again! \n")
    return playername, human, cpu


playername, human, cpu = menu()
bval = [0, 1, 2, 3, 4, 5, 6, 7, 8]
board()

# function to input human move and place it
def marking():

    j = 0
    while j == 0:
        try:
            move = int(
                input(
                    "\nEnter your move {}, type the number where you want to mark. \n".format(
                        playername
                    )
                )
            )
        except ValueError:
            print("\nWrong Input!!! Try Again \n")
            continue
        if move >= 0 and move < 9:
            if bval[move] != "X" and bval[move] != "O":
                bval[move] = human
                j = 1
            else:
                print("Position already taken, try different position!")
        else:
            print("\nWrong Input!!! Try Again \n")
    board()


# to check after and before every move if anyone wins
def checkwin():
    wld = [
        [0, 1, 2],
        [3, 4, 5],
        [6, 7, 8],
        [0, 3, 6],
        [1, 4, 7],
        [2, 5, 8],
        [0, 4, 8],
        [6, 4, 2],
    ]
    hc = [human, cpu]
    win = False
    winner = ""
    for seq in wld:
        for play in hc:
            x = 0
            for block in seq:
                if bval[block] == play:
                    x = x + 1
                else:
                    continue
            if x == 3:
                win = True
                winner = play
                return win, winner
            else:
                continue
    return win, winner


# for cpu to read the board, compare values, ck represents a board with numbers which add up to 15 [8,3,4,1,5,9,6,7,2] represents [0,1,2,3,4,5,6,7,8] positions in vertical, horizontal or diognals,
# checks if human or cpu can win in the next move, returns the positions to play for cpu in such condition.
# there are 2 functions for this because of a bug.
def cpucheckh():
    check = [
        [0, 1, 2],
        [3, 4, 5],
        [6, 7, 8],
        [0, 3, 6],
        [1, 4, 7],
        [2, 5, 8],
        [0, 4, 8],
        [6, 4, 2],
    ]
    ck = [
        [8, 3, 4],
        [1, 5, 9],
        [6, 7, 2],
        [8, 1, 6],
        [3, 5, 7],
        [4, 9, 2],
        [8, 5, 2],
        [6, 5, 4],
    ]
    hs = 0
    for x in range(8):
        h = 0
        kh = 0
        for y in range(3):
            if bval[check[x][y]] == human:
                h = h + 1
                kh = ck[x][y] + kh
            elif bval[check[x][y]] == cpu:
                h = 0
                break
            else:
                continue
        if h == 2:
            hs = 15 - kh
            return h, hs
        else:
            continue
    return 0, 0


def cpucheckc():
    check = [
        [0, 1, 2],
        [3, 4, 5],
        [6, 7, 8],
        [0, 3, 6],
        [1, 4, 7],
        [2, 5, 8],
        [0, 4, 8],
        [6, 4, 2],
    ]
    ck = [
        [8, 3, 4],
        [1, 5, 9],
        [6, 7, 2],
        [8, 1, 6],
        [3, 5, 7],
        [4, 9, 2],
        [8, 5, 2],
        [6, 5, 4],
    ]
    cs = 0
    for x in range(8):
        c = 0
        kc = 0
        for y in range(3):
            if bval[check[x][y]] == cpu:
                c = c + 1
                kc = ck[x][y] + kc
            elif bval[check[x][y]] == human:
                c = 0
                break
            else:
                continue
        if c == 2:
            cs = 15 - kc
            return c, cs
        else:
            continue
    return c, cs


# function for cpu to mark it's fav position
def cpumarking(term, a, b):
    print("\nMy move\n")
    indexs = [8, 3, 4, 1, 5, 9, 6, 7, 2]
    h, hs = cpucheckh()
    c, cs = cpucheckc()
    if c == 2:
        index = indexs.index(cs)
        if bval[index] != human and bval[index] != cpu:
            bval[index] = cpu
            board()
            return a, b
    elif h == 2:
        index = indexs.index(hs)
        if bval[index] != human and bval[index] != cpu:
            bval[index] = cpu
            board()
            return a, b

    # even term for X ,odd for O
    if term == 0 or term == 7:
        a = cpu07()
    elif term == 1:
        a = cpu1()
    elif term == 2:
        a, b = cpu2(a)
    elif term == 3 or term == 5:
        a, b = cpu35(a)
    elif term == 4:
        a = cpu4(a)
    elif term == 6:
        a, b = cpu6(a, b)
    else:
        a = cpurand()
    board()
    return a, b


# functions which mark for respective term accordingly
def cpu07():
    arr = [0, 2, 6, 8]
    for a in range(4):
        if bval[arr[a]] != human and bval[arr[a]] != cpu:
            bval[arr[a]] = cpu
            return a
        else:
            continue
    a, b = cpurand()
    return a


def cpu1():
    if bval[4] != human and bval[4] != cpu:
        bval[4] = cpu
        return 0
    else:
        a = cpu07()
        return a


def cpu2(a):
    arr = [[6, 2], [8, 0], [0, 8], [6, 2]]
    for b in range(2):
        if bval[arr[a][b]] != human and bval[arr[a][b]] != cpu:
            bval[arr[a][b]] = cpu
            return a, b
        else:
            continue
    a, b = cpurand()
    return a, b


def cpu35(a):
    arr = [[7, 5], [3, 7], [1, 5], [3, 1]]
    if bval[4] == cpu:
        for c in range(4):
            for b in range(2):
                if bval[arr[c][b]] != human and bval[arr[c][b]] != cpu:
                    bval[arr[c][b]] = cpu
                    return c, b
                else:
                    continue
    else:
        for b in range(2):
            if bval[arr[a][b]] != human and bval[arr[a][b]] != cpu:
                bval[arr[a][b]] = cpu
                return a, b
            else:
                continue
    a, b = cpurand()
    return a, b


def cpu4(a):
    arr = [8, 6, 2, 0]
    if bval[arr[a]] != human and bval[arr[a]] != cpu:
        bval[arr[a]] = cpu
        return a
    else:
        a, b = cpurand()
        return a


def cpu6(a, b):
    arr = [[3, 1], [5, 1], [3, 7], [7, 5]]
    if bval[4] != human and bval[4] != cpu:
        bval[4] = cpu
        return a, b
    elif bval[arr[a][b]] != human and bval[arr[a][b]] != cpu:
        bval[arr[a][b]] = cpu
        return a, b
    else:
        a, b = cpurand()
        return a, b


# a function to mark randomly if the conditions are not fav for cpu
def cpurand():
    rand = 0
    while rand == 0:
        arr = random.randint(0, 8)
        if bval[arr] != human and bval[arr] != cpu:
            bval[arr] = cpu
            return 0, 0
        else:
            continue


# loop to play the game, marking() and cpumarking(), check for winner at start of each loop
ai = 0
bi = 0
for i in range(9):
    win, winner = checkwin()
    if win:
        break
    elif human == "X" and i % 2 == 0:
        marking()
    elif human == "O" and i % 2 != 0:
        marking()
    else:
        ai, bi = cpumarking(i, ai, bi)
win, winner = checkwin()
if win:
    print("{} won!!\n".format(winner))
    if winner == human:
        board()
        print("This time you won {}, you won't be lucky next time.".format(playername))
    else:
        board()
        print("Know your place human, I am a superior being.")
# for the draw
if win == False and i == 8:
    print("It is a draw. You were a worthy challenger {}".format(playername))

File: tic_tac_toe.R
#mr tic tac toe
mrttt <- c("       XXXXXXXXXXXXXXXXXXXXXX",
           "    XXXXXXXXXXXXXXXXXXXXXXXXXXXXX", 
           "  XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
           " XXXXXXXXXXXXXXXXXX         XXXXXXXX", 
           "XXXXXXXXXXXXXXXX              XXXXXXX",
           "XXXXXXXXXXXXX                   XXXXX", 
           " XXX     _________ _________     XXX      ",
           "  XX    I  _xxxxx I xxxxx_  I    XX        ", 
           " ( X----I         I         I----X )           ",
           "( +I    I      00 I 00      I    I+ )", 
           " ( I    I    __0  I  0__    I    I )      /\\    /\\    |\"\"\")  -------  -------  -------", 
           "  (I    I______ /   \\_______I    I)      /  \\  /  \\   |___)     |   __   |   __   |", 
           "   I           ( ___ )           I      /    \\/    \\  |   \\ *   |        |        |", 
           "   I    _  :::::::::::::::  _    I           ",
           "    \\    \\___ ::::::::: ___/    /          ", 
           "     \\_      \\_________/      _/",
           "       \\        \\___/        /", 
           "         \\                 /",
           "          |\\             /|", 
           "          |  \\_________/  |")

# values of board and function to print the board
b_val <- c(1,2,3,4,5,6,7,8,9)

board <- function(){
  cat("\n       |       |\n")
  cat(paste("  ",b_val[1], "  |  ",b_val[2],"  |  ",b_val[3]))
  cat('\n_______|_______|_____')
  cat("\n       |       |\n")
  cat(paste("  ",b_val[4], "  |  ",b_val[5],"  |  ",b_val[6]))
  cat('\n_______|_______|_____')
  cat("\n       |       |\n")
  cat(paste("  ",b_val[7], "  |  ",b_val[8],"  |  ",b_val[9]))
  cat("\n       |       |\n")
}

#function to input the values from user
input_value <- function(string){
  if (interactive()) {
    con <- stdin()
  } else {
    con <- "stdin"
  }
  cat(string)
  symbol <- readLines(con = con,n = 1)
  return(symbol)
}

# make the menu input name and X/O choice give the other choice to cpu
menus <- function(){
  cat(mrttt, sep = "\n")
  cat("WELCOME PLAYER! I am Mr. Tic Tac Toe \nShall we play a game? \n")
  playername <- input_value("What is your name challenger?\n")
  while (T) {
    xno <- input_value(c(playername,", Select your mark: X or O \n"))
    xno <- toupper(xno)
    if (xno != "X" & xno != "O") {
      cat("\nWrong input! Try Again!\n")
      next
    }else if (xno == "X") {
      human <- "X"
      cpu <- "O"
      break
    }else if (xno == "O") {
      human <- "O"
      cpu <- "X"
      break
    }
    
  }
  return(list(playername = playername, human = human, cpu = cpu))
}
menu_val <- menus()
playername <- menu_val$playername
human <- menu_val$human
cpu <- menu_val$cpu
board()

# function to input human move and mark it on the board
marking <- function(){
  while (T) {
    move = input_value(c("\nEnter your move",playername,
                         ", type the number where you want to mark(1-9). \n"))
    move <- tryCatch(as.numeric(move),
                     error = function(e){
                       FALSE
                     },
                     warning = function(w){
                       FALSE
                     })
    
    if (!move | is.na(move)) { # when input move is invalid or empty
      cat("\nWrong Input!! Try Again \n")
    }else if (is.numeric(move)) {
      if (move >= 1 & move < 10) {
        if (b_val[move] != 'X' & b_val[move] != 'O') {
          b_val[move] <<- human
          break
        }else{
          cat("\nPosition already taken, try different position!\n")
        }
      }else{
        cat("\nWrong Input!! Enter a number from 1-9 \n")
      }
    }
  }
  board()
}

# to check after and before every move if anyone wins
# adds 1 to x for continous symbols in winning_vectors,
# Example: X|X|X will add 3, i.e. win; X|X|O add 2, no win; X|O|X will add 1, 
# and after reading O move to next vector  
checkwin <- function(){
  winning_vector <- list(c(1,2,3),c(4,5,6),c(7,8,9),c(1,4,7),
                         c(2,5,8),c(3,6,9),c(1,5,9),c(7,5,3))
  symbol_vec <- c(human, cpu)
  win <- FALSE
  winner <- ''
  for (win_seq in winning_vector) {
    for (play in symbol_vec) {
      win_sum <- 0
      for (block in win_seq) {
        if (b_val[block] == play) {
          win_sum <- win_sum + 1
        } else next
      }
      if (win_sum == 3) { 
        win <- T
        winner <- play
        return(c(win, winner))
      }else next
    }
  } 
  return(c(win, winner))
}

# function cpucheck() for cpu to read the board, compare values, 
# magic_board represents a board with numbers which add up to 15 [8,3,4,1,5,9,6,7,2] represents 
# [1,2,3,4,5,6,7,8,9] positions in vertical, horizontal or diagonals,
#
#   8    |   3  |  4  +=15
# ________________________
#   1   |   5  |  9  += 15
# ________________________
#   6  |   7  |  2  +=15
#  _______________________
# +=15| +=15 |+=15

# Similarly it adds up to 15 for diagonals
# checks if human or cpu can win in the next move, i.e.
# if hum = 2, there are 2 continuous symbols, win possible in next move
# returns the positions to play for cpu in such condition, i.e.
# 15 - sum of magic board sum, example if sum is 13, the next move needs 2 more,
# so according to magic board with reference to board is 9th position.
cpucheck <- function(player){
  check <- rbind(c(1,2,3),c(4,5,6),c(7,8,9),c(1,4,7),
                 c(2,5,8),c(3,6,9),c(1,5,9),c(7,5,3))
  magic_board <- rbind(c(8,3,4),c(1,5,9),c(6,7,2),c(8,1,6),
                       c(3,5,7),c(4,9,2),c(8,5,2),c(6,5,4))
  mgk_scr <- 0
  if (player == 'X')
  {
    enemy <- 'O'
  } else enemy <- 'X'
  
  for (x in 1:8) { # each vector 
    win_sum <- 0
    mgk_brd_sum <- 0
    for (y in 1:3) { # each symbol
      if (b_val[check[x,y]] == player) {
        win_sum <- win_sum + 1
        mgk_brd_sum <- magic_board[x,y] + mgk_brd_sum
        # if the continuous marks is broken by enemy mark set win sum to 0
      }else if (b_val[check[x,y]] == enemy) { 
        win_sum <- 0
        break
      }else next
    }
    # if next move may end up as win or loose
    if (win_sum == 2) { 
      mgk_scr <- 15 - mgk_brd_sum
      return(c(win_sum,mgk_scr))
    }else next
  }
  return(c(0,0))
}

#function for cpu to mark it's fav position
cpumarking <- function(term,fav1,fav2) {
  cat("\nMy move\n")
  indexs <- c(8,3,4,1,5,9,6,7,2)
  hum_scr <- cpucheck(human)
  cpu_scr <- cpucheck(cpu)
  
  # priority check if win or loose is possible in next move
  if (cpu_scr[1] == 2) {
    index <- which(indexs == cpu_scr[2])[[1]]
    if (b_val[index] != human & b_val[index] != cpu) {
      b_val[index] <<- cpu
      board()
      return(c(fav1,fav2)) 
    }
  }else if (hum_scr[1] == 2) {
    index <- which(indexs == hum_scr[2])[[1]]
    if (b_val[index] !=  human & b_val[index] !=  cpu) {
      b_val[index] <<- cpu
      board()
      return(c(fav1,fav2))
    }
  }
  # even term for X ,odd for O
  if (term == 1 | term == 8) {
    fav1 <- cpu1_8()
  }else if (term == 2) {
    fav1 <- cpu2() 
  }else if (term == 3) {
    fav <- cpu3(fav1)
    fav1 <- fav[1]
    fav2 <- fav[2]
  }else if (term == 4 | term == 6) {
    fav <- cpu4_6(fav1)
    fav1 <- fav[1]
    fav2 <- fav[2]
  }else if (term == 5) {
    fav1 <- cpu5(fav1) 
  }else if (term == 7) {
    fav <- cpu7(fav1,fav2)
    fav1 <- fav[1]
    fav2 <- fav[2]
  }else {
    fav <- cpu_random()
    fav1 <- fav[1]
    fav2 <- fav[2]
    cat()
    
  }
  board()
  return(c(fav1,fav2))
}

# functions which mark for respective term accordingly
# the odd numbers are for 'X' and even numbers are for 'O' 
# as more moves are placed, 
cpu1_8 <- function() {
  arr <- c(1,3,7,9)
  for (fav_pos in 1:4) {
    if (b_val[arr[fav_pos]] != human & b_val[arr[fav_pos]] !=  cpu) {
      b_val[arr[fav_pos]] <<- cpu
      return(fav_pos)     
    }else next
  }
# return 1 if fav position was already taken
  random <- cpu_random()
  return(random[1])
}

# position 5(center) is fav in 2nd move, this is for 'O'
cpu2 <- function(){ 
  if (b_val[5] != human & b_val[5] !=  cpu) {
    b_val[5] <<- cpu
    return(1)
  }else {
    fav1 <- cpu1_8() 
    return(fav1)
  }
}

# input the fav. vector and choose between the 2 best pos.
cpu3 <- function(fav1){ 
  arr <- rbind(c(7,3),c(9,1),c(1,9),c(7,3))
  for (fav2 in 1:2) {
    if (b_val[arr[fav1,fav2]] != human & b_val[arr[fav1,fav2]] != cpu) {
      b_val[arr[fav1,fav2]] <<- cpu
      return(c(fav1,fav2))
    }else next
  }
# if fav. positions are already filled return a random pos.
  random <- cpu_random() 
  return(random)
}

# if the cpu has the centre vs if it doesn't returns different fav. positions.
cpu4_6 <- function(fav1){
  arr <- rbind(c(8,6),c(4,8),c(2,6),c(4,2))
  if (b_val[5] == cpu) { 
    for (fav3 in 1:4) {
      for (fav2 in 1:2) {
        if (b_val[arr[fav3,fav2]] != human & b_val[arr[fav3,fav2]] != cpu) {
          b_val[arr[fav3,fav2]] <<- cpu
          return(c(fav3,fav2)) 
        }else next
      }
    }
  }else{
    for (fav2 in 1:2) { 
      if (b_val[arr[fav1,fav2]] != human & b_val[arr[fav1,fav2]] != cpu) {
        b_val[arr[fav1,fav2]] <<- cpu
        return(c(fav1,fav2))
      }else next
    }
  }
  random <- cpu_random()
  return(random)
}

cpu5 <- function(fav1){
  arr <- c(9,7,3,1)
  if (b_val[arr[fav1]] != human & b_val[arr[fav1]] != cpu) {
    b_val[arr[fav1]] <<- cpu
    return(fav1)
  }else {
    random <- cpu_random()
    return(random[1])
  }
}

cpu7 <- function(fav1,fav2){
  arr <- rbind(c(4,2),c(6,2),c(4,8),c(8,6))
  if (b_val[5] != human & b_val[5] != cpu) {
    b_val[5] <<- cpu
    return(c(fav1,fav2))
  }else if (b_val[arr[fav1,fav2]] != human & b_val[arr[fav1,fav2]] != cpu) {
    b_val[arr[fav1,fav2]] <<- cpu
    return(c(fav1,fav2))
  }else {
    random <- cpu_random()
    return(random)
  }
}

# a function to mark randomly if the conditions are not fav for cpu
cpu_random <- function(){
  arr <- sample(1:9, 9, replace = FALSE)
  for (fav in arr) {
    if (b_val[fav] != human & b_val[fav] != cpu) {
      b_val[fav] <<- cpu
      return(c(1,1))
    }else next
  }
}

win_loose <- function(win)
{
  if (win[1]) {
    cat(paste(win[2], " won!!\n"))
  }
  if (win[2] == human) {
    cat(paste("This time you won" ,playername ,", you won't be lucky next time."))
  }else if (win[2] == cpu) {
    cat(paste("Good game, well played" ,playername ,",but you can't beat me in your lifetime!"))
  }else if (win[1] == FALSE) {
    cat("It is a draw. You were a worthy challenger! Next time I will win.", playername)
  }
}
# to play the game, mark by marking() and cpumarking() and check for winner at start of each loop
# global var to be used in the functions
fav1i <- 1
fav2i <- 1
play <- function(turn)
{ win <- checkwin()
  if (win[1]) {
    return(0)
  }else if (human == 'X' & turn %% 2 != 0) {
    marking()
  }else if (human == 'O' & turn %% 2 == 0) {
    marking()
  }else{
    mark <- cpumarking(turn,fav1i,fav2i)
    fav1i <<- mark[1]
    fav2i <<- mark[2]
  }
  
  return(c(fav1i, fav2i))
}

play_var <- sapply(1:9, play)
win <- checkwin()
win_loose(win)
